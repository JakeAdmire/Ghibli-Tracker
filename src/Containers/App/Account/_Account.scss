.Account {
  width: 100%; height: 100%;
  @include flexBox(column, center, center);


  form {
    width: 500px; height: 500px;
    font-family: 'Raleway';
    @include flexBox(column, initial, initial);

    div {
      @include flexBox(row, center, space-between);

      button { width: 48%; }
      .disabled { background: $accent-hover; }
    }
  }

  label {
    margin-top: 20px;
    @include customFont(24px, 600, none, $accent-color);
  }

  input {
    box-sizing: border-box;
    padding: 10px;

    &::placeholder {
      color: white;
      opacity: 0.7;
      transition: $ease;
    }

    &:focus {
      &::placeholder { opacity: 0; }
    }
  }

  input,
  button {
    height: 50px;
    border-radius: 15px;
    border: none;
    transition: $ease;
    background: $accent-color;
    @include customFont(24px, 600, none, white);
  }

  button {
    &:hover { 
      cursor: pointer;
      background: $accent-hover;
    }
    &:active { opacity: 0.5; }
  }

  .submit { margin-top: 30px; }
}