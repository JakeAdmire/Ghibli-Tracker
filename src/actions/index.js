export const toggleFavorited = (id) => ({
  type: 'TOGGLE_FAVORITED',
  id
})

export const setFilter = (filter) => ({
  type: 'SET_FILTER',
  filter
})

export const toggleInfo = (id) => ({
  type: 'TOGGLE_INFO',
  id
})

export const addFilms = (films) => ({
  type: 'ADD_FILMS',
  films
})

export const addUser = (name, email, password) => ({
  type: 'ADD_USER',
  name,
  email,
  password
})

export const loginUser = (email, password) => ({
  type: 'LOGIN_USER',
  email,
  password
})