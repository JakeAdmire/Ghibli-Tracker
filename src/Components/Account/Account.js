import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addUser, loginUser } from '../../actions';

export class Account extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      email: '',
      password: ''
    }
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { name, email, password } = this.state;
    const { value } = e.target;
    const loginData = {
      email: email.toLowerCase(),
      password: password
    } 
    const signupData = {
      name: name,
      email: email.toLowerCase(),
      password: password
    } 
    if (value === 'Log In') {
      this.validateUser('/api/users', loginData);
    }
    if (value === 'Sign Up') {
      this.createUser('/api/users/new', signupData);
    }
  }

  async validateUser(url = '', loginData = {}) {
    try {
      const response = await fetch(`http://localhost:3000${url}`, {
        method: 'POST',
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(loginData),
      })
      const result = await response.json();
      this.props.loginUser(result.data.id, loginData.name);
    } catch(error) {
        throw new Error('incorrect email/password');
      }
  }

  async createUser(url = '', signupData = {}) {
    try {
      const response = await fetch(`http://localhost:3000${url}`, {
        method: 'POST',
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(signupData), 
      })
      const data = await response.json();
      this.props.loginUser(data.id, signupData.name); 
    } catch(error) {
      throw new Error('user already exists');
    }
  }

  render() {
    return (
      <form>
        <label htmlFor='name'>Name:</label>
        <input type='text' name='name' value={this.state.name} onChange={this.handleChange} />
        <label htmlFor='email'>Email:</label>
        <input type='text' name='email' value={this.state.email} onChange={this.handleChange} />
        <label htmlFor='password'>Password:</label>
        <input type='text' name='password' value={this.state.password} onChange={this.handleChange} />
        <button value="Log In" onClick={this.handleSubmit}>Log In</button>
        <button value="Sign Up" onClick={this.handleSubmit}>Sign Up</button>
      </form>
    )
  }
}


const mapDispatchToProps = (dispatch) => ({
  loginUser: (id, name) => dispatch(loginUser(id, name)),
})

export default connect(null, mapDispatchToProps)(Account);