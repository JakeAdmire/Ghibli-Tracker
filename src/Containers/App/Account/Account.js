import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import { loginUser } from '../../../actions';

export class Account extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      email: '',
      password: '',
      loggedIn: false,
      errorMessage: '',
      toggleSignUp: false
    }
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { value } = e.target;
    const formData = this.state;
    formData.email = formData.email.toLowerCase();
    if (value === 'Log In') {
      this.validateUser('/api/users', formData);
    }
    if (value === 'Sign Up') {
      this.createUser('/api/users/new', formData);
    }
  }

  async validateUser(url = '', loginData = {}) {
    try {
      const results = await this.fetchUser(url, loginData);
      this.props.loginUser(results.data.id, results.data.name);
      this.setState({loggedIn: true})
    } catch(error) {
      this.setState({errorMessage: 'Sorry, you entered an incorrect email or password'})
      }
  }

  async createUser(url = '', signupData = {}) {
    try {
      const results = await this.fetchUser(url, signupData);
      this.props.loginUser(results.id, signupData.name); 
    } catch(error) {
      throw new Error('user already exists');
    }
  }

  async fetchUser(url, data) {
    try {
      const response = await fetch(`http://localhost:3000${url}`, {
        method: 'POST',
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(data), 
      })
      const results = await response.json();
      return results;
    } catch(error) {
      throw new Error('Request unsuccessful')
    }
  }

  render() {
    return (
      <div className='Account'>
        {
          !this.state.toggleSignUp ? 
            <form>
              <div>
                <button onClick={() => this.setState({toggleSignUp: true})}>Sign In</button>
                <button className='disabled'>Create Account</button>
              </div>
              <label htmlFor='name'>Name:</label>
              <input placeholder='Kim..' type='text' name='name' value={this.state.name} onChange={this.handleChange} required/>
              <label htmlFor='email'>Email:</label>
              <input placeholder='ex. kim@mail.com..' type='text' name='email' value={this.state.email} onChange={this.handleChange} required/>
              <label htmlFor='password'>Password:</label>
              <input placeholder='ex. floomples..' type='text' name='password' value={this.state.password} onChange={this.handleChange} required/>
              <button className='submit' value="Sign Up" onClick={this.handleSubmit}>SIGN UP</button>
            </form> :
            <form >
              <div>
                <button className='disabled'>Sign In</button>
                <button onClick={() => this.setState({toggleSignUp: false})}>Create Account</button>
              </div>
              <label htmlFor='email'>Email:</label>
              <input placeholder='ex. kim@mail.com..' type='text' name='email' value={this.state.email} onChange={this.handleChange} required/>
              <label htmlFor='password'>Password:</label>
              <input placeholder='ex. floomples..' type='text' name='password' value={this.state.password} onChange={this.handleChange} required/>
              <button className='submit' value="Log In" onClick={this.handleSubmit}>LOG IN</button>
            </form>  
        }
        {
          this.state.loggedIn && <Redirect to='/' />
        }
        {
          this.state.errorMessage ? (
            <div>{this.state.errorMessage}</div>
          ) : (
            <div>Please sign in or create an account</div>
          )
        }
      
      </div>
    )
  }
}


export const mapDispatchToProps = (dispatch) => ({
  loginUser: (id, name) => dispatch(loginUser(id, name)),
})

export default connect(null, mapDispatchToProps)(Account);

Account.propTypes = {
    loginUser: PropTypes.func
}