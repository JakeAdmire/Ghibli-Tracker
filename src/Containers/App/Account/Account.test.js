import React from 'react';
import { shallow } from 'enzyme';
import { Account } from './Account';
import { loginUser } from '../../../actions';
import { mapDispatchToProps } from './Account';

describe('Account', () => {
  let wrapper;
  let mockEvent;

  beforeEach(() => {
    wrapper=shallow(<Account loginUser={loginUser}/>)
  })

  describe('Account', () => {
    it ('should have the correct default state', () => {
      expect(wrapper.state()).toEqual({
        name: '',
        email: '',
        password: '',
        loggedIn: false,
        errorMessage: ''
      })
    })

    it('should setState for name when handleChange is invoked', () => {
      mockEvent = {target: {name: 'name', value: 'Kim'}}
      wrapper.instance().handleChange(mockEvent)
      expect(wrapper.state()).toEqual({
        name: 'Kim',
        email: '',
        password: '',
        loggedIn: false,
        errorMessage: ''
      })
    })

    it('validateUser should call fetchUser with the correct url and data if Log In clicked', () => {
      const mockUrl = 'www.users.com'
      const mockFormData = {
        email: 'kim@gmail.com',
        password: 'kim'
      }
      wrapper.instance().validateUser = jest.fn()
      wrapper.instance().fetchUser = jest.fn()
      wrapper.instance().validateUser(mockUrl, mockFormData)
      wrapper.instance().fetchUser(mockUrl, mockFormData)
      expect(wrapper.instance().fetchUser).toHaveBeenCalledWith(mockUrl, mockFormData)
    })

    it('createUser should call fetchUser with the correct url and data if Sign Up clicked', () => {
      const mockUrl = 'www.newusers.com'
      const mockFormData = {
        name: 'Kim',
        email: 'kim@gmail.com',
        password: 'kim'
      }
      wrapper.instance().createUser = jest.fn()
      wrapper.instance().fetchUser = jest.fn()
      wrapper.instance().createUser(mockUrl, mockFormData)
      wrapper.instance().fetchUser(mockUrl, mockFormData)
      expect(wrapper.instance().fetchUser).toHaveBeenCalledWith(mockUrl, mockFormData)
    })

    it('fetchUser should return expected data if LogIn is clicked', async () => {
      const mockUrl = 'www.getusers.com'
      const mockResults = {
        status: "success",
        data: [
          {
          id: 1,
          name: "Taylor",
          password: "password",
          email: "tman2272@aol.com"
          }
          ],
          message: "Retrieved All Users"
        }
        wrapper.instance().fetchUser = jest.fn().mockImplementation(() => Promise.resolve({
          ok: true,
          json: () => Promise.resolve(mockResults)
        }))

        const response = await wrapper.instance().fetchUser(mockUrl)
        const result = await response.json()
        expect(result).toEqual(mockResults)
    })

    it('fetchUser should throw error if response fails', async () => {
      const mockUrl = 'www.getusers.com'
      wrapper.instance().fetchUser = jest.fn().mockImplementationOnce(() => Promise.resolve({
        ok: false
      }))
      try {
        await wrapper.instance().fetchUser(mockUrl)
      } catch(error) {
        expect(error.message).toBe('Request unsuccessful')
      }
    })

    it('should match the snapshot', () => {
      expect(wrapper.debug()).toMatchSnapshot()
    })

  })

  describe('mapDispatchToProps', () => {
    it('should provide a method to use to dispatch an action creator', () => {
      const mockId = 1
      const mockName = 'Jake'
      const mockDispatch = jest.fn()
      const actionToDispatch = loginUser(mockId, mockName)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.loginUser(mockId, mockName)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })


})