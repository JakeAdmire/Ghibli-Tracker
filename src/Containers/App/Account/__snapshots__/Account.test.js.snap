// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Account should match the snapshot 1`] = `
"<div>
  <form>
    <label htmlFor=\\"email\\">
      Email:
    </label>
    <input type=\\"text\\" name=\\"email\\" value=\\"\\" onChange={[Function]} required={true} />
    <label htmlFor=\\"password\\">
      Password:
    </label>
    <input type=\\"text\\" name=\\"password\\" value=\\"\\" onChange={[Function]} required={true} />
    <button value=\\"Log In\\" onClick={[Function]}>
      Log In
    </button>
  </form>
  <form>
    <label htmlFor=\\"name\\">
      Name:
    </label>
    <input type=\\"text\\" name=\\"name\\" value=\\"\\" onChange={[Function]} required={true} />
    <label htmlFor=\\"email\\">
      Email:
    </label>
    <input type=\\"text\\" name=\\"email\\" value=\\"\\" onChange={[Function]} required={true} />
    <label htmlFor=\\"password\\">
      Password:
    </label>
    <input type=\\"text\\" name=\\"password\\" value=\\"\\" onChange={[Function]} required={true} />
    <button value=\\"Sign Up\\" onClick={[Function]}>
      Sign Up
    </button>
  </form>
  <div>
    Please sign in or create an account
  </div>
</div>"
`;
